<launch>
  <arg name="input_depth_image" default="depth/image" />
  <arg name="input_rgb_image" default="rgb/image_rect_color" />
  <arg name="input_point" default="depth/points" />
  <arg name="input_info" default="rgb/camera_info" />

  <node pkg="jsk_pcl_ros" type="border_estimator" name="border_estimator"
	output="screen">
    <remap from="~input" to="$(arg input_point)" />
    <remap from="~input_camera_info" to="$(arg input_info)" />
  </node>

  <!-- edge detector -->
  <node pkg="jsk_perception" type="edge_detector" name="edge">
    <remap from="image" to="$(arg input_rgb_image)" />
  </node>

  <node pkg="image_view" type="image_view" name="edge_view" >
    <remap from="image" to="edge/image" />
  </node>
  
  <!-- relay image to pcl_nodelet name space -->
  <node pkg="topic_tools" type="relay" name="edge_relay"
        args="in out">
    <remap from="in" to="edge/image"/>
    <remap from="out" to="pcl_nodelet/edge/image" />
  </node>

  <group ns="pcl_nodelet">
    <node pkg="nodelet" type="nodelet"
          name="mask2pi"
          args="standalone jsk_pcl/MaskImageToPointIndices"
          output="screen" clear_params="true">
      <remap from="~input" to="edge/image" />
      <rosparam>
        filter_field_name: z
        filter_limit_min: 0.0
        filter_limit_max: 5.0
        leaf_size: 0.01
      </rosparam>
    </node>
  </group>

  <node pkg="nodelet" type="nodelet" name="indices_add"
	args="standalone jsk_pcl/AddPointIndices">
    <remap from="~input/src1" to="pcl_nodelet/mask2pi/output" />
    <remap from="~input/src2" to="border_estimator/output_border_indices" />
    <rosparam>
      approximate_sync: true
      use_indices: true
      negative: false
      max_queue_size: 5000
    </rosparam>
  </node>

  <node pkg="nodelet" type="nodelet" name="added"
	args="standalone pcl/ExtractIndices">
    <remap from="~input" to="$(arg input_point)" />
    <remap from="~indices" to="pcl_nodelet/mask2pi/output" />
    <rosparam>
      approximate_sync: true
      use_indices: true
      negative: false
      max_queue_size: 100
    </rosparam>
  </node>

  <include file="$(find jsk_pcl_ros)/launch/tracking.launch">
    <arg name="input_point" value="added/output" />
    <!-- <arg name="input_point" default="/pcl_nodelet/candidate_points/output" /> -->
    <arg name="voxel_grid_filter_limit_min" value="0.0"/>
    <arg name="voxel_grid_filter_limit_max" value="5.0"/>
    <arg name="voxel_grid_filter_field_name" value="true"/>
    <arg name="voxel_grid_leaf_size" value="0.02"/>
  </include>

</launch>
